// react-document/lib/Document.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var parts_1 = require("./parts");
exports.documentPropTypes = {
    htmlProps: PropTypes.object,
    bodyProps: PropTypes.object,
    children: PropTypes.string,
    title: PropTypes.string,
    viewport: PropTypes.string,
    head: PropTypes.node,
    footer: PropTypes.node,
    css: PropTypes.arrayOf(PropTypes.string),
    js: PropTypes.arrayOf(PropTypes.string),
    appMountId: PropTypes.string,
    app: PropTypes.string,
    window: PropTypes.object,
};
var Document = (function (_super) {
    __extends(Document, _super);
    function Document() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Document.prototype.getChildContext = function () {
        var props = Object.assign({}, this.props);
        if (props.children && !props.app) {
            props.app = props.children;
            delete props.children;
        }
        return {
            documentProps: props,
        };
    };
    Document.prototype.render = function () {
        return (React.createElement(parts_1.Html, null,
            React.createElement(parts_1.Head, null),
            React.createElement(parts_1.Body, null,
                React.createElement(parts_1.App, null),
                React.createElement(parts_1.Footer, null))));
    };
    return Document;
}(React.Component));
Document.childContextTypes = {
    documentProps: PropTypes.shape(exports.documentPropTypes).isRequired,
};
Document.propTypes = exports.documentPropTypes;
Document.defaultProps = {
    app: '',
    appMountId: 'app',
    viewport: 'width=device-width, initial-scale=1',
};
exports.default = Document;
