// react-document/lib/parts.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var serialize = require("serialize-javascript");
var Document_1 = require("./Document");
var DocumentPart = (function (_super) {
    __extends(DocumentPart, _super);
    function DocumentPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DocumentPart;
}(React.Component));
DocumentPart.contextTypes = {
    documentProps: PropTypes.shape(Document_1.documentPropTypes).isRequired
};
exports.DocumentPart = DocumentPart;
var Html = (function (_super) {
    __extends(Html, _super);
    function Html() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Html.prototype.render = function () {
        return (React.createElement("html", __assign({}, this.context.documentProps.htmlProps, this.props)));
    };
    return Html;
}(DocumentPart));
exports.Html = Html;
var Head = (function (_super) {
    __extends(Head, _super);
    function Head() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Head.prototype.render = function () {
        var _a = __assign({}, this.context.documentProps, this.props), title = _a.title, viewport = _a.viewport, head = _a.head, css = _a.css, js = _a.js;
        return (React.createElement("head", null,
            React.createElement("meta", { charSet: "utf-8" }),
            title && React.createElement("title", null, title),
            React.createElement("meta", { name: "viewport", content: viewport }),
            head && head.map(function (element, i) { return React.cloneElement(element, { key: i }); }),
            css && css.map(function (file) { return React.createElement("link", { key: file, href: file, rel: "stylesheet" }); }),
            js && js.map(function (file) { return React.createElement("script", { key: file, src: file, defer: true }); }),
            this.props.children));
    };
    return Head;
}(DocumentPart));
exports.Head = Head;
var Body = (function (_super) {
    __extends(Body, _super);
    function Body() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Body.prototype.render = function () {
        return (React.createElement("body", __assign({}, this.context.documentProps.bodyProps, this.props)));
    };
    return Body;
}(DocumentPart));
exports.Body = Body;
var App = (function (_super) {
    __extends(App, _super);
    function App() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    App.prototype.render = function () {
        var id = this.props.id || this.context.documentProps.appMountId;
        var content = this.props.children || this.context.documentProps.app;
        return (React.createElement("div", { id: id, dangerouslySetInnerHTML: { __html: content || '' } }));
    };
    return App;
}(DocumentPart));
exports.App = App;
var Footer = (function (_super) {
    __extends(Footer, _super);
    function Footer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Footer.prototype.render = function () {
        var _a = __assign({}, this.context.documentProps, this.props), window = _a.window, footer = _a.footer;
        var children = this.props.children;
        if (!window && !footer && !children) {
            return null;
        }
        return (React.createElement("div", null,
            window && (React.createElement("script", { dangerouslySetInnerHTML: { __html: Object.keys(window).map(function (key) {
                        return "window['" + key + "'] = " + serialize(window[key]) + ";";
                    }).join('\n')
                } })),
            footer &&
                footer.map(function (element, i) { return React.cloneElement(element, { key: i }); }),
            children));
    };
    return Footer;
}(DocumentPart));
exports.Footer = Footer;
